openssl/src/stack.rs

```diff
@@ -0,0 +1,275 @@
+use std::marker;
+use std::ops;
+use std::mem;
+use std::iter;
+
+use libc::c_void;
+
+use ffi;
+use opaque::Opaque;
+
+/// Trait implemented by stackable types. This must *only* be
+/// implemented on opaque types that can be directly casted into their
+/// `CType`.
+pub trait Stackable {
+    /// Underlying C type for this object.
+    type CType;
+    /// C stack type for this element. Generally called
+    /// `stack_st_{ELEMENT_TYPE}`, normally hidden by the
+    /// `STACK_OF(ELEMENT_TYPE)` macro in the OpenSSL API.
+    type StackType;
+    /// When the stack is owned this destructor is used to free each
+    /// element.
+    fn destructor() -> unsafe extern fn(*mut Self::CType);
+}
+
+/// Reference to a stack of `T`. No destructors will be run when this
+/// is dropped. This is akin to a Rust slice.
+pub struct StackRef<T: Stackable> {
+    stack: Opaque,
+    phantom: marker::PhantomData<T>,
+}
+
+impl<T: Stackable> StackRef<T> {
+    pub unsafe fn from_ptr<'a>(stack: *const T::StackType)
+                               -> &'a StackRef<T> {
+        &*(stack as *const _)
+    }
+
+    pub unsafe fn from_ptr_mut<'a>(stack: *mut T::StackType)
+                                   -> &'a mut StackRef<T> {
+        &mut *(stack as *mut _)
+    }
+
+    pub fn as_ptr(&self) -> *mut T::StackType {
+        &self.stack as *const _ as *mut _
+    }
+
+    /// OpenSSL stack types are just a (kinda) typesafe wrapper around
+    /// a `_STACK` object. We can therefore safely cast it and access
+    /// the `_STACK` members without having to worry about the real
+    /// layout of `T::StackType`.
+    ///
+    /// If that sounds unsafe then keep in mind that's exactly how the
+    /// OpenSSL 1.1.0 new C stack code works.
+    fn _as_stack(&self) -> *const ffi::_STACK {
+        self.as_ptr() as *const _
+    }
+
+    /// Returns the number of items in the stack
+    pub fn len(&self) -> usize {
+        self._len()
+    }
+
+    #[cfg(ossl10x)]
+    fn _len(&self) -> usize {
+        unsafe { (*self._as_stack()).num as usize }
+    }
+
+    #[cfg(ossl110)]
+    fn _len(&self) -> usize {
+        unsafe { ffi::OPENSSL_sk_num(self.stack as *const _) as usize }
+    }
+
+    pub fn iter(&self) -> Iter<T> {
+        Iter {
+            stack: self,
+            pos: 0,
+        }
+    }
+
+    pub fn iter_mut(&mut self) -> IterMut<T> {
+        IterMut {
+            stack: self,
+            pos: 0,
+        }
+    }
+
+    /// Returns a reference to the element at the given index in the
+    /// stack or `None` if the index is out of bounds
+    pub fn get(&self, idx: usize) -> Option<&T> {
```

Comment:
Whenever we're returning `&T`, we should probably be returning `&Ref<T>`, and same for `&mut T`.